### ISSUE TITLE
How to have individulal learning rates for each neuron in some layers?

### ISSUE THREAD

**Issue by arashno** [torch/nn#466]
Hi all.
I am trying to train a deep neural networks with 10 layers.
I want to set individual learning rates for each neuron in some layers(not all!).
I know that it is possible to set individual learning rates for all the neurons via  optim/sgd.lua.
But I don't know how to do this just some layers, not all.
Thanks

**Comment by nkoumchatzky** [comment]
Hi arashno. A solution would be to group the parameters per-layer in a
table, such as groupedParams =
{{weightLayer1,biasLayer1}, {weightLayer2,biasLayer1}, ..., {weightLayer10,
biasLayer10}}, same for your gradParameters in a groupedGradParams, then
loop over this table like this for sgd:

for i=1,#groupedParams do
   optim.sgd(function(x) return _, groupedGradParams[i] end,
groupedParams[i], config[i], state[i])
end

where config is a table of optimization hyper-params (such as your learning
rates). In order to group the params, you'll have to do it yourself though.

Hope this helps,

Nico


2015-11-10 1:40 GMT+01:00 arashno <notifications@github.com>:

**Comment by arashno** [comment]
Thanks for your reply.
So basically you suggest calling SGD for each layer separately.
Two questions come to my mind:
1- would this method gives the same result as calling SGD for all layers?
2- how efficient it would be?
Thanks

**Comment by soumith** [comment]
@arashno for SGD it should be the same result. It should be a tad bit less efficient, but in the overall scheme of things, you wont notice the slowdown.

**Comment by egaebel** [comment]
I am also trying to perform layer-wise training of a neural network in torch, but I'm having some slight problems. I've trained this network in the normal fashion and now I'm training it layer-wise to compare results and make sure I'm doing this right, but the performance is way off which is leading me to believe I'm made a mistake.

for t, v in ipairs(indices) do
    local inputs = trainingDataLoader.data:index(1, v)
    targets:copy(trainingDataLoader.labels:index(1, v))

    gradParameters:zero()

    -- Run network forward and backward
    local outputs = net:forward(inputs)
    local f = criterion:forward(outputs, targets)
    local df_do = criterion:backward(outputs, targets)
    net:backward(inputs, df_do)
    confusion:batchAdd(outputs, targets)

    -- Perform layer-wise optimization on the weights
    for i = 1, #layerParameters do


        if layerParameters[i]:nDimension() == 0 then
            print("Skipping layer!\n\n")
        else
            local feval = function(x)
                return _, layerGradParameters[i]
            end
            optim.sgd(feval, layerParameters[i], layerOptimConfig[i])
        end
    end
end

More detailed code here:
http://paste.ofcode.org/33jBvHr7THsvZVaCEVFHm8p

Thanks!

**Comment by dasguptar** [comment]
@egaebel , did you finally find what was wrong with your snippet?
I was going to try layer-wise learning rates, and came across this post!

**Comment by egaebel** [comment]
Yes I did! I do not remember exactly what I changed, and I may look at this a bit deeper later but this was the code that worked for me. Layer-wise training ended up being a bit of a hassle and didn't really improve anything for my use-case, so I abandoned it. But I hope it proves useful for you!

`-- Setup table for optimizer parameters
local memoryLength = opt.batchSize
optimConfig = {
    learningRate = opt.learningRate,
    weightDecay = opt.weightDecay,
    momentum = opt.momentum,
    learningRateDecay = opt.learningRateDecay
}
local function copyTable(mytable)
    newtable = {}
    for k, v in pairs(mytable) do
        newtable[k] = v
    end
    return newtable
end

------------------------------------------------------------------
-- The basic goal of this separate main.lua file is the following
-- Use sgd to train on each layer of the net
-- this requires me to get separate...EVERYTHING for each layer
-- separate parameters, separate gradParameters, etc.
------------------------------------------------------------------

-- Get NN parameters and set up confusion matrix
parameters, gradParameters = net:getParameters()
-- Get layer-wise parameters and gradParameters
local layerOptimConfig = {}
local layerParameters = {}
local layerGradParameters = {}
for i = 1, net:size() do
    layer = net:get(i)
    local parameters, gradParameters = layer:getParameters()
    table.insert(layerParameters, parameters)
    table.insert(layerGradParameters, gradParameters)
    -- Create copies of optimConfig so they can be altered on the fly later
    table.insert(layerOptimConfig, copyTable(optimConfig))
end

local confusion = optim.ConfusionMatrix(trainingDataLoader.classes)

-- Train the net, layer by layer, using adaptive learning rates
local function train()
    if not epoch then
        epoch = 1
    else
        epoch = epoch + 1
    end
    collectgarbage()
    net:training()
    -- Setup mini-batches
    local indices = torch.randperm(trainingDataLoader.data:size(1)):long():split(opt.batchSize)
    indices[#indices] = nil
    local targets = torch.Tensor(opt.batchSize)
    if not EXCLUDE_CUDA_FLAG then
        targets = targets:cuda()
    end

    print('Training Beginning....')
    print('Training set size: ', trainingDataLoader.data:size(1))
    local tic = torch.tic()
    for t, v in ipairs(indices) do
        local inputs = trainingDataLoader.data:index(1, v)
        targets:copy(trainingDataLoader.labels:index(1, v))

        -- Zero (I'm actually not sure why this has to be done 
        --  on each layerGradParameter..there's something messed up with 
        --  references I think...)
        gradParameters:zero()
        for i = 1, #layerGradParameters do
            layerGradParameters[i]:zero()
        end

        -- Run network forward and backward
        local outputs = net:forward(inputs)
        local f = criterion:forward(outputs, targets)
        local df_do = criterion:backward(outputs, targets)
        net:backward(inputs, df_do)
        confusion:batchAdd(outputs, targets)

        -- Perform layer-wise optimization on the weights
        for i = 1, #layerParameters do
            -- If this layer has parameters to be optimized
            if layerParameters[i]:nDimension() ~= 0 then
                if epoch > 1 then
                    -- Allocate space on first runs, zero on subsequent runs
                    if not layerOptimConfig[i].learningRates then
                        layerOptimConfig[i].learningRates = torch.zeros(layerGradParameters[i]:size(1), 1)
                    else
                        layerOptimConfig[i].learningRates:zero()
                    end

                    -- ADJUST LEARNING RATES DYNAMICALLY HERE
                    --
                    --
                    --
                    --
                end
                -- Perform optimization
                local feval = function(x)
                    return _, layerGradParameters[i]
                end
                optim.sgd(feval, layerParameters[i], layerOptimConfig[i])
            end
        end
    end
    confusion:updateValids()
    print(('Train accuracy: ' .. c.cyan'%.2f' .. ' %%\t time: %.2f s'):format(
            confusion.totalValid * 100, torch.toc(tic)))
    trainingAccuracy = confusion.totalValid * 100
    
    print('------------------------------------------------')
    print('Confusion Matrix:')
    print(confusion)
    confusion:zero()
    print('------------------------------------------------')
end`

**Comment by alykhantejani** [comment]
You can do this in a simple way with `optim.sgd` as it accepts a `learningRates` parameter in the `config`.

**Comment by gpleiss** [comment]
For anyone who's interested, I just published a package for layer-wise learning rates.
https://github.com/gpleiss/nnlr

It's code I've used for some of my own projects, similar to @alykhantejani 's idea

(It's very beta, so feedback appreciated!)
